workflow-trace-viewer/src/jvmMain/kotlin/com/squareup/workflow1/traceviewer/WorkflowJsonParser.kt

```
@@ -0,0 +1,29 @@
+package com.squareup.workflow1.traceviewer
+
+import com.squareup.moshi.JsonDataException
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
+import java.io.IOException
+
+/**
+ * Parses a JSON string into [WorkflowNode] with Moshi adapters.
+ *
+ * All the caught exceptions should be handled by the caller, and appropriate UI feedback should be
+ * provided to user.
+ */
+public fun parseTrace(
```

JSON parsing _can_ be long running. This would be better as a `suspend fun` that wraps this work in a `withContext(Dispatchers.IO)` block. Then, instead of having a `jsonString` being used as state in your `WorkflowContent` Composable, you'd have a nullable `WorkflowNode`.

That would make `WorkflowContent` look something like this:

```kotlin
@Composable
private fun WorkflowContent(file: PlatformFile?) {
  var rootNode by remember { mutableStateOf<WorkflowNode?>(null) }
  LaunchedEffect(file) {
     if(file != null) { 
       rootNode = parseTrace(file?.readString()) // in the future, a try-catch here to switch to an error state!
    }
  }

  if (rootNode != null) {
    DrawWorkflowTree(root)
  } else {
    Text("Empty data or failed to parse data") // TODO: proper handling of error
  }
}
```

For the future, when you get to the point where you might wanna show an error on parse failures, you'd instead just let this throw and expect it to be caught in the `LaunchedEffect` you call this from so that the consumer could switch to some error state. But that's looking ahead so don't stress that now!

---

Working on this in another branch!